src/main/scala/com/fulcrumgenomics/bam/ErrorRateByReadPosition.scala

```diff
@@ -0,0 +1,280 @@
+        (rec.getMappingQuality < minMappingQuality)
+    }
+    .foreach { rec =>
+      counter.accept(rec)
+      progress.record(rec)
+    }
+
+    val path = PathUtil.pathTo(output.getOrElse(PathUtil.removeExtension(input)) + ErrorRateByReadPositionMetric.FileExtension)
+    Metric.write[ErrorRateByReadPositionMetric](metrics=counter.metrics, path=path)
+
+    in.safelyClose()
+  }
+}
+
+object ErrorRateByReadPositionMetric {
+  val FileExtension = ".error_rate_by_read_position.txt"
+}
+
+/** Gives the error rate and total number of bases observed for each position within the read.
+  *
+  * @param read_number the read number (ex. 0 for fragments, 1 for read one of a pair, 2 for read two of a pair)
+  * @param position the read position (1-based).
+  * @param error_rate the error rate at the given position n the read
+  * @param bases_total the total number of bases observed at this position.
+  */
+case class ErrorRateByReadPositionMetric
+( read_number: Int,
+  position: Int,
+  error_rate: Double,
+  bases_total: Long
+) extends Metric
+
+private object ErrorRateByReadPositionCounter {
+  def apply(ref: Option[PathToFasta], includeInsertions: Boolean, minBaseQuality: Int): ErrorRateByReadPositionCounter = {
+    ref match {
+      case Some(r) => new ErrorRateByReadPositionWithReferenceCounter(ref=r, includeInsertions=includeInsertions, minBaseQuality=minBaseQuality)
+      case None    => new ErrorRateByReadPositionWithMdCounter(includeInsertions=includeInsertions, minBaseQuality=minBaseQuality)
+    }
+  }
+}
+
+/** Base class for computing the error rate by position.
+  *
+  * Sub-classes should implement the [updateErrors] method to extract the read bases and qualities, and reference bases.
+  * They should then call [incrementIfValid] or [increment] to update the counters for each read offset.
+  *
+  * @param minBaseQuality the minimum base quality to allow.
+  */
+private abstract class ErrorRateByReadPositionCounter(includeInsertions: Boolean, minBaseQuality: Int = 0) {
+  private val errorsByReadOffset = List.tabulate(3)(x => new NumericCounter[Int]())
+  private val totalByReadOffset  = List.tabulate(3)(x => new NumericCounter[Int]())
+
+  /** Gets the list of metrics sorted by read number and then position. */
+  def metrics: Seq[ErrorRateByReadPositionMetric] = {
+    if (totalByReadOffset.forall(_.isEmpty)) return Seq.empty
+    // create the metrics
+    errorsByReadOffset.zip(totalByReadOffset).zipWithIndex.filter {
+      case ((errCounter, totalCounter), idx) => totalCounter.nonEmpty
+    }.flatMap { case ((errCounter, totalCounter), idx) =>
+      val masOffset = totalCounter.map(_._1).max
+      Range.inclusive(0, masOffset).map { offset =>
+        new ErrorRateByReadPositionMetric(
+          read_number = idx,
+          position    = offset + 1,
+          error_rate  = errCounter.countOf(offset) / totalCounter.countOf(offset).toDouble,
```

Comment:
I _think_ there are pathological situations where this could cause a divide by zero error.  E.g. if the read is 100 bases long, but everything at cycle 99 (for example) had Q < `minBaseQuality`, you could have zero in the totalCounter.